localstack-core/localstack/services/lambda_/provider.py

```diff
@@ -2141,37 +2155,53 @@ def update_event_source_mapping_v2(
         event_source_mapping = old_event_source_mapping | request_data
 
-        if function_name_or_arn:
-            # if the FunctionName field was present, update the FunctionArn of the EventSourceMapping
-            account_id, region = api_utils.get_account_and_region(function_name_or_arn, context)
-            function_name, qualifier = api_utils.get_name_and_qualifier(
-                function_name_or_arn, None, context
-            )
-            event_source_mapping["FunctionArn"] = api_utils.qualified_lambda_arn(
-                function_name, qualifier, account_id, region
-            )
-
         temp_params = {}  # values only set for the returned response, not saved internally (e.g. transient state)
 
+        # Validate the newly updated ESM object. We ignore the output here since we only care whether an Exception is raised.
+        function_arn, _, _ = self.validate_event_source_mapping(context, event_source_mapping)
+
+        # remove the FunctionName field
+        event_source_mapping.pop("FunctionName", None)
+
+        if function_arn:
+            event_source_mapping["FunctionArn"] = function_arn
+
         esm_worker = self.esm_workers[uuid]
         # Only apply update if the desired state differs
         enabled = request.get("Enabled")
         if enabled is not None:
             if enabled and old_event_source_mapping["State"] != EsmState.ENABLED:
-                esm_worker.start()
                 event_source_mapping["State"] = EsmState.ENABLING
             # TODO: What happens when trying to update during an update or failed state?!
             elif not enabled and old_event_source_mapping["State"] == EsmState.ENABLED:
-                esm_worker.stop()
                 event_source_mapping["State"] = EsmState.DISABLING
+        else:
+            event_source_mapping["State"] = EsmState.UPDATING
+
+        # To ensure parity, certain responses need to be immediately returned
+        temp_params["State"] = event_source_mapping["State"]
 
         state.event_source_mappings[uuid] = event_source_mapping
+
+        # TODO: Currently, we re-create the entire ESM worker. Look into approach with better performance.
+        function_version = get_function_version_from_arn(function_arn)
+        function_role = function_version.config.role
+        worker_factory = EsmWorkerFactory(
+            event_source_mapping, function_role, request.get("Enabled", esm_worker.enabled)
+        )
+
+        # Get a new ESM worker object but do not active it, since the factory holds all logic for creating new worker from configuration.
+        updated_esm_worker = worker_factory.get_esm_worker()
+        self.esm_workers[uuid] = updated_esm_worker
+
+        # We should stop() the worker since the delete() will remove the ESM from the state mapping.
+        esm_worker.stop()
+        # This will either create an EsmWorker in the CREATING state if enabled. Otherwise, the DISABLING state is set.
```

Comment:
This comment is a bit misleading, it will create an EsmWorker with the internal state of `CREATING`, but the state in the store depends on the actual action we are taking. In the followups, we should discuss the state management of ESM workers, I am not a fan at all with the current approach, but good job making the best out of it!